### PR TITLE
Add --code flag for collecting code context.

### PR DESCRIPTION
**Pull request by rnemes**
## Bug / Requirement Description
Include the 1 liner of the testcase assertion (top-level from testcase) into the report.

## Solution description
Added ``--code`` flag to collect code context for the assertions. Code context one-liner will be displayed on the web UI if enabled.
Note that file path information is no longer collected by default. To collect file path information, enable code context.

## Checklist:
- [ ] Test
- [ ] Example (both test_plan.py and .rst)
- [ ] Documentation (API)
- [x] News fragment present for release notes
- [x] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [x] For new cmdline arg: documentation

### CODE REVIEW

File: testplan/web_ui/testing/src/Common/utils.js [hunk 1]
```diff
@@ -447,3 +447,17 @@ export const truncateString = (
   
   return `${startPortion}...${endPortion}`;
 };
+
+/**
+ * Extracts the "workspace" part of the path
+ * @param {String} path
+ */
+export const getWorkspacePath = (path) => {
+  const srcIndex = path.indexOf('/src/');
```
Comment (rnemes) [inline testplan/web_ui/testing/src/Common/utils.js:456]: This only works in linux

File: tests/unit/testplan/testing/multitest/test_result.py [hunk 2]
```diff
@@ -43,17 +43,29 @@ def intermediary(result, description=None):
...
+
+    # Functions to test marking, with relative position of the assertions
+    functions = [
+        (test_group_marking, 6),
+        (helper, 1),
+        (intermediary, 2),
+    ]
+    current_function = inspect.currentframe().f_code.co_name
+    for function, relative_position in functions:
+        # To prevent infinite recursion
```
Comment (Pyifan) [inline tests/unit/testplan/testing/multitest/test_result.py:62]: The modification to this testcase doesn't look right to me.

Let's discuss offline.

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: testplan/testing/multitest/base.py [hunk 3]
```diff
@@ -1158,6 +1158,9 @@ def _run_testcase(
                 ),
```
Comment (Pyifan) [inline testplan/testing/multitest/base.py:1158]: when --code is not specified, we don't need to call report_target on testcase either. Every report_target call evokes inspect and is expensive.

File: testplan/testing/multitest/base.py [hunk 4]
```diff
@@ -1158,6 +1158,9 @@ def _run_testcase(
...
             )
 
+        if getattr(self.cfg, "collect_code_context", False):
+            testcase = collect_code_context(func=testcase)
```
Comment (Pyifan) [inline testplan/testing/multitest/base.py:1162]: This doesn't make sense, the decorator sets a threading.local variable, why are do calling it on every testcase?

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: testplan/testing/multitest/base.py [hunk 5]
```diff
@@ -1158,6 +1158,9 @@ def _run_testcase(
                 ),
             )
 
+        if getattr(self.cfg, "collect_code_context", False):
```
[the remark this answers is not in the document]
Comment (rnemes) [reply testplan/testing/multitest/base.py:1161]: Several tests under "unit/testplan/testing/multitest/test_multitest" fail if I don't use getattr

File: tests/unit/testplan/testing/multitest/test_result.py (hunk 2 above)
Comment (rnemes) [reply tests/unit/testplan/testing/multitest/test_result.py]: Reverted back to the previous style as we discussed

File: testplan/testing/multitest/base.py (hunk 3 above)
Comment (rnemes) [reply testplan/testing/multitest/base.py]: modified: report_target is now no-op when --code is not set

File: testplan/testing/base.py [hunk 6]
```diff
@@ -901,6 +901,14 @@ def driver_info(self) -> bool:
...
         return self.cfg.driver_info
 
+    @property
+    def collect_code_context(self) -> bool:
+        """
+        Collecting the file path, line number and code context of the assertions
+        if enabled.
+        """
+        return getattr(self.cfg, "collect_code_context", False)
+
```
[the remark this answers is not in the document]
Comment (zhenyu-ms) [reply testplan/testing/base.py:911]: might need it for hooks & maybe result call under pytest wrapper

File: testplan/testing/multitest/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhenyu-ms) [reply testplan/testing/multitest/base.py:1151]: the getattr call should be inside that `collect_code_context` property in `testplan/testing/base.py`

[the remark this answers is not in the document]
Comment (rnemes) [reply testplan/testing/multitest/base.py]: > the getattr call should be inside that `collect_code_context` property in `testplan/testing/base.py`

Good catch, makes sense!

File: testplan/testing/multitest/base.py (hunk 4 above)
Comment (rnemes) [reply testplan/testing/multitest/base.py]: I have refactored the solution and added using Result for storing the flag as we discussed it online.

File: testplan/testing/result.py [hunk 7]
```diff
@@ -1383,8 +1392,11 @@ def __exit__(self, exc_type, exc_value, traceback):
             self.description,
             self.category,
         )
-        assertion.file_path = os.path.abspath(caller_frame[1])
-        assertion.line_no = caller_frame[2]
+
+        if caller_frame:
+            assertion.file_path = os.path.abspath(caller_frame[1])
+            assertion.line_no = caller_frame[2]
+            # assertion.code_context = caller_frame.code_context[0].strip()
```
Comment (Pyifan) [inline testplan/testing/result.py:1399]: why is this commented out?

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: testplan/testing/result.py (hunk 7 above)
Comment (rnemes) [reply testplan/testing/result.py:1399]: very good question, not intentionally

File: testplan/web_ui/testing/src/Common/utils.js (hunk 1 above)
Comment (zhenyu-ms) [reply testplan/web_ui/testing/src/Common/utils.js:456]: 🤔

Comment (rnemes) [reply testplan/web_ui/testing/src/Common/utils.js:456]: (Discussed this offline.) This is a request from management, the full path will be visible in the popup.

**Review by Pyifan — APPROVED** [review]

**Pull request merged**